projects/plugins/jetpack/_inc/client/main.jsx

```diff
@@ -199,6 +200,87 @@ class Main extends React.Component {
 	}
 
 	renderMainContent = route => {
+		/**
+		 * There are two conditions (groups of conditions, really) where we would want to
+		 * show the partner connection screen.
+		 *
+		 * 1. The site is not yet connected to WPCOM, but has the jetpack_partner_coupon
+		 * option set in the database (this.props.partnerCoupon in redux). This is likely a
+		 * partner-user who has just arrived here from a CTA within a partner's dashboard
+		 * or other ecosystem.
+		 *
+		 * 2. The site is already connected to WPCOM, but the jetpack_partner_coupon option
+		 * is still set in the database. This means the user connected their site, but never
+		 * redeemed the coupon. If this is the case, we don't want to override the dashboard
+		 * or at a glance pages with the redemption screen. Instead, we'll catch a URL
+		 * parameter that JITMs will set (showCouponRedemption=true), and show the screen only
+		 * when the user came from a a JITM.
+		 */
+		if ( this.props.partnerCoupon ) {
+			const forceShow = new URLSearchParams( window.location.search ).get( 'showCouponRedemption' );
+			const partnerCoupon = this.props.partnerCoupon;
+
+			if ( ! this.props.isSiteConnected || forceShow ) {
+				this.partnerCouponHandleClick = () => {
+					window.location.href = getRedirectUrl( 'jetpack-plugin-partner-coupon-checkout', {
+						path: partnerCoupon.product.slug,
+						site: this.props.siteRawUrl,
+						query: `coupon=${ partnerCoupon.coupon_code }`,
+					} );
+				};
+
+				return (
+					<ConnectScreen
+						apiNonce={ this.props.apiNonce }
+						registrationNonce={ this.props.registrationNonce }
+						apiRoot={ this.props.apiRoot }
+						images={ [ '/images/products/illustration-backup.png' ] }
+						assetBaseUrl={ this.props.pluginBaseUrl }
+						title={ sprintf(
+							/* translators: %s: Jetpack partner name. */
+							__( 'Welcome to Jetpack %s traveler!', 'jetpack' ),
+							partnerCoupon.partner
+						) }
+						buttonLabel={ sprintf(
+							/* translators: %s: Name of a Jetpack product. */
+							__( 'Set up & redeem %s', 'jetpack' ),
+							partnerCoupon.product.title
+						) }
+						redirectUri={ `admin.php?page=jetpack&partnerCoupon=${ partnerCoupon.coupon_code }` }
+						connectionStatus={ this.props.connectionStatus }
+					>
+						<p>
+							{ sprintf(
+								/* translators: %s: Name of a Jetpack product. */
+								__(
+									'Redeem your coupon and get started with %s for free the first year!',
+									'jetpack'
+								),
+								partnerCoupon.product.title
+							) }
+						</p>
+						<ul>
+							{ partnerCoupon.product.features.map( ( feature, key ) => (
+								<li className="jp-recommendations-product-purchased__feature" key={ key }>
+									{ feature }
+								</li>
+							) ) }
+						</ul>
+						{ this.props.connectionStatus.hasConnectedOwner && (
+							<ActionButton
```

Comment:
@rcoll the ToS text when this button is displayed is lacking underscores in links (also, the links aren't blue like the design, but that's for both "connected" and "pre-connected" and I don't feel like we have to change this to be different from the original pre-connection screen.